### PR TITLE
Update multiformats, deps, examples, docs

### PR DESCRIPTION
**Pull request by rvagg**
First .. this is very heavy-handed of me so please tell me off where I'm stepping over bounds of style etc.! I'm happy to scale back, or have you take my commits and scale them back, if you have strong opinions about some of the things I've adjusted in the process of this PR.

I spent a bit of time familiarising myself with the workflow here and how it's all supposed to work. In the process I fleshed out the examples a bit more and got the signing example from the README fully working (it seemed to be partial). The signing work also prompted https://github.com/ceramicnetwork/js-dag-jose-utils/pull/1 to get some minor utilities in place. I'm not an expert in JOSE so you may have to correct me if I'm getting the workflow wrong!

Examples are now included as separate files, separate ones for encryption with and without IPFS - so demonstrating best practice now for bridging into the js-IPFS world from the multiformats world, but also demonstrating stand-alone IPLD usage without needing the js-IPFS pieces. A separate example for signing seemed appropriate since it's such a different workflow.

Dependencies are trimmed down quite a bit thanks to the affordances of the new stack.

I've copied the current multiformats codec pattern: https://github.com/multiformats/js-multiformats/blob/1b133a87f8c85bb8f1f088ca78821d87a3a0988a/src/codecs/interface.ts#L21, where we've removed `default` exports for codecs (we've basically ditched default exports across the stack, it's too hard to support across different bundlers and resolvers). So it exports `{name, code, encode, decode}` now but also exports a `prepare` function. This same pattern is used in https://github.com/ipld/js-dag-pb where a similar thing is happening—input objects are being transformed such that a round-trip leaves you with a different object, and the forms are very strict on input, which is all unlike dag-cbor and dag-json which are flexible. So instead, `prepare()` can optionally get you to the same form of object that you would have if you were to round-trip through an encode and decode (optionally because I didn't want to break the pattern of letting `encode()` still accept the `string` form). Also in support of this round-trippishness I added `link` onto the `prepare()`d form of `DagJWS`, so it should just always be there.

Custom Jest resolver is pending fix being tracked @ https://github.com/facebook/jest/issues/9771

Jest `"testEnvironment": "node"` is set pending jsdom getting proper `TextEncoder` and `TextDecoder` support like the rest of the JS world, tracked @ https://github.com/facebook/jest/issues/9983

### CODE REVIEW

File: example-ipld.mjs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+/** @type {Map<string, Uint8Array>} */
+const store = new Map()
+
+async function symmetric () {
+  // Encrypt and store a payload using a secret key
+  const storeEncrypted = async (payload, key) => {
+    const dirEncrypter = xc20pDirEncrypter(key)
+    // prepares a cleartext object to be encrypted in a JWE
+    const cleartext = await prepareCleartext(secretz)
```
Comment (oed) [inline example-ipld.mjs:30]:
```suggestion
    const cleartext = await prepareCleartext(payload)
```

File: example-signing-ipld.mjs [hunk 2]
```diff
@@ -0,0 +1,88 @@
+// JWT & utilities
+import {
+  ES256KSigner,
+  createJWS,
+  verifyJWS
+} from 'did-jwt'
```
Comment (oed) [inline example-signing-ipld.mjs:6]: We've been discussing renaming this library to `did-jose` since it does much more than just JWTs :)

File: src/encryption.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { fromBase64url, toBase64url } from './utils'
 
 interface JWERecipient {
-  encrypted_key?: string // eslint-disable-line @typescript-eslint/camelcase
+  encrypted_key?: string // eslint-disable-line
```
Comment (oed) [inline src/encryption.ts:4]: Why did you remove the rule that was being ignored here? Now I guess all rules are ignored?

File: src/index.ts [hunk 4]
```diff
@@ -2,66 +2,106 @@ import signing, { DagJWS, EncodedJWS } from './signing'
...
+    'iv' in jose &&
+    typeof jose.iv === 'string' &&
+    'protected' in jose &&
+    typeof jose.protected === 'string' &&
+    'tag' in jose &&
+    typeof jose.tag === 'string'
+  )
+}
+
+export function prepare(jose: DagJWS | DagJWE | string): DagJWS | DagJWE {
```
Comment (oed) [inline src/index.ts:62]: This function should probably be called `compactToGeneral` for clarity (converts compact encoding to general encoding).

File: src/index.ts [hunk 5]
```diff
@@ -2,66 +2,106 @@ import signing, { DagJWS, EncodedJWS } from './signing'
...
+    return jose
+  }
+  throw new Error('Not a valid unencoded JOSE object')
+}
+
+export function encode(obj: DagJWS | DagJWE | string): ByteView<EncodedJWS | EncodedJWE> {
   if (typeof obj === 'string') {
-    encodedJose = stringToJose(obj)
-  } else if (isJWS(obj)) {
+    obj = prepare(obj)
```
Comment (oed) [inline src/index.ts:80]: Right, this happens here so that's why I was confused why you called `prepare` in the example in the readme.

File: test/index.test.ts [hunk 6]
```diff
@@ -1,18 +1,18 @@
...
 import ipldInMem from 'ipld-in-memory'
-import CID from 'cids'
-import { EllipticSigner } from 'did-jwt'
+import { CID } from 'multiformats/cid'
+import { ES256KSigner } from 'did-jwt'
 
 describe('dag-jose codec', () => {
   describe('DagJWS', () => {
-    it('Encode compact jws', () => {
+    it.only('Encode compact jws', () => {
```
Comment (oed) [inline test/index.test.ts:15]: Should not be `.only`?

**Review by oed — COMMENTED** [review]
This looks really awesome! Thanks for doing a deep dive and creating all of these examples 🙏  
Left a few comments, but nothing major.

You may also be interesting in having a look at [js-did](https://github.com/ceramicnetwork/js-did) to get a sense of how we use this codec together with the DID standard.

Ping @ukstv, you might be curious about this code base :)

File: README.md [hunk 7]
```diff
@@ -1,118 +1,228 @@
...
-const privkey = // hex private key
-const signer = new EllipticSigner(privkey)
+const signer = ES256KSigner(privkey)
+// arbitrary data to DAG-CBOR encode, we get a:
+// { cid:CID, linkedBlock: Uint8Array }
+const payloadBlock = await encodePayload(payload)
+// sign the CID as a JWS using our signer
+const jws = await createJWS(toJWSPayload(payloadBlock), signer)
+// convert our JWS to a DagJWS for IPLD encoidng
+const dagJWS = dagJose.prepare(jws)
```
[the remark this answers is not in the document]
Comment (rvagg) [reply README.md:63]: yes .. but .. there's a problem of symmetry, as an example I think it'd be better to demonstrate having properly isomorphic objects, the `dagJWS` you have here is going to be identical to the `dagJWS` you'd have if it went through the encode/decode round trip. It also opens up the possibility of deprecating and removing the sloppy handling in `encode()` like we've done with dag-pb -- which requires that you give it exactly the form of object that will be present in a round-trip, the user is responsible for handing the codec a properly formed data model object, but we can give them utilities to help them get there.

if you look at the current example in the readme you see this problem potentially popping up (I ran into it when messing around with it myself). Unless the object goes through the encode/decode then you can't use it properly. There are legimitate cases where you may never encode an object, but need to have its valid data model form.

basically - best practice is to have a clean mapping of the data model form such that it's stable through an encode/decode cycle.

but this is only an example, and maybe this needs to be explained?

File: src/encryption.ts (hunk 3 above)
Comment (rvagg) [reply src/encryption.ts:4]: I was getting errors about it being missing, deprecated: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/camelcase.md

I can put it back but I think this might be a new typescript-eslint thing, and I don't really know how to control the new `naming-convention` rule inline like this -- got any suggestions?

File: test/index.test.ts (hunk 6 above)
Comment (rvagg) [reply test/index.test.ts:15]: whoops! leftover from debugging, will remove

**Comment by rvagg** [comment]
started having a look at the comments but have run out of time, will get back to this asap

File: src/encryption.ts (hunk 3 above)
Comment (oed) [reply src/encryption.ts:4]: Was mostly curious. What happens if you remove the `eslint-disable-line` entirely?

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (oed) [reply README.md:63]: So any JOSE library will (for JWS) create an object which has three proerties: `payload`, `header`, and `signature` (or some variation of this depending on the JOSE encoding). With DagJWS we add an additional property called `link` which contains a CID instance which is derived from the payload (which is a base65url CID). This link property will only be populated when retrieving the content from the dag so that we can traverse it.

The main reason I want the codec to support multiple JOSE encodings is so that you can just put the output of any JOSE library directly into the DAG without having to think more about it.

File: src/encryption.ts (hunk 3 above)
Comment (rvagg) [reply src/encryption.ts:4]: nothing! have removed it 🤞

File: README.md [hunk 8]
```diff
@@ -1,118 +1,228 @@
...
+const jwsBlock = await Block.decode({ bytes, codec: dagJose, hasher: sha256 })
+const jwsStrings = toJWSStrings(jwsBlock.value)
+// verify the signatures found in the block against our pubkey
+for (const jws of jwsStrings) {
+  const verifiedKey = verifyJWS(jws, [{ publicKeyHex: pubkey }]) // will throw if it doesn't verify
+  console.log(`Verified JWS envelope \u001b[32m${cid}\u001b[39m with public key:\n\t${verifiedKey.publicKeyHex}`)
+}
+
+const payloadCid = jwsBlock.value.link
+const payloadBytes = store.get(payloadCid.toString())
```
[the remark this answers is not in the document]
Comment (rvagg) [reply README.md:92]: have added comments, maybe too verbose, ptal

**Comment by rvagg** [comment]
@oed I've renamed `prepare()` to `toGeneral()`, commented it out in the example but documented it there. It's really useful to have a way to ensure you have the same form of object that you'd get through a round-trip so I think this is a useful export. Your call though. I was thinking about just doing a `toCompact()` to provide the two-way option and then maybe `toJWSStrings()` in dag-jose-utils could be removed, _but_ I'm not sure at all how to handle the multi-sig option. That doesn't seem to be handled here at all. `fromSplit()` only handles a single signature.

So I'm wondering if:

1. `toGeneral()` should also accept a `string[]` with `fromSplit()` - in this case I'm not sure whether `encryption.ts/fromSplit()` has an array form for `DagJWE` (if so, what does it do?)
2. add `toCompact()` which is essentially the same as `toJWSStrings()` .. and also handles the `DagJWE` case (again, not sure if there's an array form of this, although I see `.recipients` is an array so maybe that's the multiple form?).

Thoughts?

**Comment by oed** [comment]
> I was thinking about just doing a toCompact() to provide the two-way option

The compact encoding only supports a single signer/recipient, so it would be weird to return multiple compact encodings. Ideally the did-jwt library *should* support the general encoding.

1. No, as I think having an array of strings doesn't make much sense from the perspective of the JOSE spec.
2. I would not add a `toCompact()` function!

I think most of these issues stem from the **did-jwt** library not yet supporting the general encoding. If it did we wouldn't need any of these additional conversions. Rather fix things there than to add more confusing conversion functions here.

**Comment by rvagg** [comment]
> The compact encoding only supports a single signer/recipient

ok, that simplifies things then. So this has a `toGeneral()` which I guess covers the basics. IMO this is good to go, I've just updated dependencies again so we're up to date.

**Review by oed — APPROVED** [review]
Thanks for putting all of this together @rvagg!

**Comment by oed** [comment]
Published as `dag-jose@1.0.0` 🎉
